src/monitors/BytecodeProfilingMonitor.v3

```diff
@@ -0,0 +1,423 @@
+def E = MonitorRegistry.add(
+	"profile_bytecode", "Recovers source level information from bytecode execution",
+	BytecodeProfilingMonitor.new());
+
+def OUT = Trace.OUT;
+
+// the minimum number of cases contained in the "switch" statement of the dispatch
+// function. This is used during static analysis to identify the most likely wasm function
+// as dispatch.
+var MIN_DISPATCH_CASE_COUNT = 150;
+
+// the maximum size in bytes that separates 2 consecutive opcodes.
+var MAX_OPERAND_LEN = 6;
+
+class BytecodeProfilingMonitor extends Monitor {
+	def visitor = DispatchFunctionBytecodeVisitor.new(BytecodeIterator.new());
+	def fuel_count_probe = FuelCountProbe.new();
+	// the SourceProfileState of all functions that could be the dispatch function.
+	var potential_dispatch_functions: List<SourceProfileState>;
+	var all_counters: List<DynamicLoc -> Resumption>;
+
+	def configure(args: string) -> string {
+		def argMap = MonitorUtil.argMapFrom(args);
+		if (argMap.has("switch_size")) {
+			var parse_result = Ints.parseDecimal(argMap["switch_size"], 0);
+			if (parse_result.0 > 0) MIN_DISPATCH_CASE_COUNT = parse_result.1;
+			return null;
+		}
+		if (argMap.has("operand_len")) {
+			var parse_result = Ints.parseDecimal(argMap["operand_len"], 0);
+			if (parse_result.0 > 0) MAX_OPERAND_LEN = parse_result.1;
+			return null;
+		}
+		def buf = StringBuilder.new();
+		buf.put1("invalid argument(s) \"%s\"", args);
+		return buf.toString();
+	}
+
+	def onParse(module: Module) {
+		// enable global count probe.
+		Instrumentation.insertGlobalProbe(fuel_count_probe);
+
+		// detect possible dispatch functions and insert memory probes.
+		var mi = ModuleInstrumenter.new(module);
+		mi.forEachFunc(instrumentFunc);
+
+		// insert entry/exit probe at all functions.
+		mi.beforeAllFuncExecAndReturn(beforeFuncExecOrReturn(_, true), beforeFuncExecOrReturn(_, false));
+	}
+
+	def onFinish(i: Instance, r: Result) {
+		for (states = potential_dispatch_functions; states != null; states = states.tail) {
+			states.head.compactTrace();
+			states.head.reportSourceFunctionTrace();
+			states.head.reportRecoveredBytecodes();
+			states.head.hotness.report();
+		}
+	}
+
+	private def instrumentFunc(mi: ModuleInstrumenter, func: FuncDecl) {
+		visitor.reset();
+		visitor.bi.reset(func).dispatchAll(visitor);
+		if (visitor.is_dispatch_func) { // save the profile state and insert memory probes 
+			var state = SourceProfileState.new(fuel_count_probe, func.func_index,
+				StringBuilder.new().put1("func_%d", func.func_index).toString());
+			
+			potential_dispatch_functions = List.new(state, potential_dispatch_functions);
+			var memory_visitor = MemoryVisitor.new(mi.module,
+				beforeMemRead(state, visitor.opcode_read_pc, _, _, _, _), null, null);
+			memory_visitor.bi = visitor.bi.reset(func);
+			memory_visitor.bi.dispatchAll(memory_visitor);
+		}
+	}
+
+	private def beforeMemRead(profile_state: SourceProfileState, opcode_read_pc: int,
```

Comment:
It should be possible to separate the opcode read from other reads statically based on the pc and add a different probe at different locations?

Reply:
I made them separate by using 2 MemoryVisitors, saving the outermost conditional branching